Added Build & SSH keys APIs documentations (#2761)
cfengine/documentation: reference/enterprise-api-ref/build-api.markdown

---
layout: default title: Build API published: true tags: [reference, enterprise, API, build, modules]
---

> The Build API enables you to easily manage policy projects and their respective CFBS modules.

Reviewer: ```suggestion
The Build API enables you to easily manage policy projects and their respective CFEngine Build modules.
```

# Projects API

A project is a set of CFBS modules and custom files/json/policy files.

## Create project

**URI:** https://hub.cfengine.com/api/build/projects

**Method:** POST

**Parameters:**

* **repositoryUrl** *(string)*
  Git repository URL. Project will be synchronized with this repository. Supported protocols: `http`, `https`, `ssh`
  , `git`. Required. Git repository URL example: https://github.com/cfengine/build-index.git

* **branch** *(string)*
  Repository branch. Required.
* **name** *(string)*
  Project name. Required.
* **authenticationType** *(string)*
  Authentication type that will be used to get access to the repository. Allowed values: `password`, `ssh_key`.
  Required.
* **username** *(string)*
  Username for authentication to the repository. Required when authentication type is `password`.
* **password** *(string)*
  Password for authentication to the repository. Required when authentication type is `password`.
* **sshPrivateKey** *(string)*
  SSH private key for authentication to the repository. Required when authentication type is `ssh_key` and `sshKeyId` is
  not set.

> * **sshKeyId** *(integer)*

Author: yeah, should be link to the SSH keys API, I will add

> * **sshKeyId** *(integer)*

Author: hm, the link already exists:

Generated SSH private key ID by [SSH keys API][SSH keys API#Generate SSH key]

> * **sshKeyId** *(integer)*

Author: > Can we put a few details in here about how we generate the keys with the AP
yeah, if this helps I can mention that a key is in OpenSSH format and  SSH_DIGEST_ALGORITHM  is  sha512 and
SSH_PRIVATE_KEY_BITS is 4096

>they don't match the specs/requirements of our system

I don't know any specific requirements of our system

| 204 No content | Project successfully synced |
| 404 Not found | Project not found |
| 500 Internal server error | Internal server error |

##  Refresh project

Fetch upstream repository and return the current state.

**URI:** https://hub.cfengine.com/build/projects/:id/refresh

**Method:** POST

**Parameters:**

* **id** *(integer)*
  Project's ID. Required.

**Example request (curl):**

```console
curl --user <username>:<password> \
  -X POST \
  https://hub.cfengine.com/api/build/projects/2/refresh
```

**Successful response example:**

```
HTTP 200 OK
{
    "status": "ahead",
    "details": [
        "Refreshed repository for project 4 with 'git fetch'"
    ]
}
```
**Output:**

* **status**
  Project's status. Possible values: 
  - `ok` - project is up-to-date
  - `behind` - there are not pulled changes on the upstream side
  - `ahead` - there are not pushed changes on the local project side
  - `diverged` - there are not pushed and pulled changes at the same time 

**Responses:**

| HTTP response code | Description |
|--------|-------------|
| 200 Ok | Successful response  |
| 404 Not found | Project not found |
| 500 Internal server error | Internal server error |

> ## List of added to project CFBS modules

Reviewer 2: Yea that makes sense

                    "latest": false
                },
                "1.0.1": {
                    "date": "Nov 1, 2021",
                    "latest": true
                }
            },
            "latest": true,
            "ts_vector": "'autorun':1A,3B 'enable':2B 'functionality':4B"
        }
    ],
    "meta": {
        "count": 1,
        "page": 1,
        "timestamp": 1657097484,
        "total": 1
    }
}
```

**Responses:**

| HTTP response code | Description |
|--------|-------------|
| 200 Ok | Successful response  |
| 500 Internal server error | Internal server error |

# Update available CFBS modules list

Modules will be received from the official CFEngine Build modules catalogue https://build.cfengine.com

**URI:** https://hub.cfengine.com/api/build/modules

**Method:** POST

**Example request (curl):**

```console
curl --user <username>:<password> \
  -X POST \
  https://hub.cfengine.com/api/build/modules
```

**Successful response example:**

```
204 No content
```

**Responses:**

| HTTP response code | Description |
|--------|-------------|
| 204 No content | Modules list successfully updated |
| 500 Internal server error | Internal server error |

> # Get CFBS module by name

Reviewer: ```suggestion
# Get CFEngine build module by name
```

    "readme": "<h1 id=\"enable-autorun\">Enable autorun</h1>\n<p>Simple module to enable autorun functionality, using def.json.</p>\n",
    "description": "Enable autorun functionality",
    "version": "1.0.1",
    "author": {
        "url": "https://github.com/olehermanse",
        "name": "Ole Herman Schumacher Elgesem",
        "image": "https://avatars.githubusercontent.com/u/4048546?v=4"
    },
    "updated": "2021-11-03 00:00:00+00",
    "downloads": 1837,
    "repo": "https://github.com/cfengine/modules",
    "documentation": null,
    "website": null,
    "subdirectory": "management/autorun",
    "commit": "c3b7329b240cf7ad062a0a64ee8b607af2cb912a",
    "dependencies": "[]",
    "tags": "[\"supported\", \"management\"]",
    "versions": {
        "1.0.0": {
            "date": "Oct 26, 2021",
            "latest": false
        },
        "1.0.1": {
            "date": "Nov 1, 2021",
            "latest": true
        }
    },
    "latest": true
}
```

**Responses:**

| HTTP response code | Description |
|--------|-------------|
| 200 Ok | Successful response  |
| 404 Not found | Module not found  |
| 500 Internal server error | Internal server error |

> # Get specific version of CFBS module by name

Reviewer: ```suggestion
# Get specific version of a CFEngine Build module by name
```

---
layout: default
title: Build API
published: true
tags: [reference, enterprise, API, build, modules]
---

The Build API enables you to easily manage policy projects and their respective CFEngine Build modules.

# Projects API

A project is a set of CFEngine Build modules and custom files/json/policy files.

## Create project

**URI:** https://hub.cfengine.com/api/build/projects

**Method:** POST

**Parameters:**

* **repositoryUrl** *(string)*
  Git repository URL. Project will be synchronized with this repository. Supported protocols: `http`, `https`, `ssh`

>   , `git`. Required. Git repository URL example: https://github.com/cfengine/build-index.git

Reviewer: Rather than use build-index as an example of a project url we could throw up an exaple cfbs built repo and point to that, or just point to a completely fake one.

* **branch** *(string)*
  Repository branch. Required.
* **name** *(string)*
  Project name. Required.
* **authenticationType** *(string)*
  Authentication type that will be used to get access to the repository. Allowed values: `password`, `ssh_key`.
  Required.
* **username** *(string)*
  Username for authentication to the repository. Required when authentication type is `password`.
* **password** *(string)*
  Password for authentication to the repository. Required when authentication type is `password`.
* **sshPrivateKey** *(string)*
  SSH private key for authentication to the repository. Required when authentication type is `ssh_key` and `sshKeyId` is
  not set.
* **sshKeyId** *(integer)*
  Generated SSH private key ID by [SSH keys API][SSH keys API#Generate SSH key] for authentication to the repository.
  Required when authentication type is `ssh_key` and `sshPrivateKey` is not set.

> 

Reviewer 3: maybe add 

```
** Note: the SSH key is expected to be in openssh(rfc4716) format as generated by [SSH Keys API][SSH keys API#Generate SSH key] or a command line like

ssh-keygen -t rsa-sha2-512 -b 4096 -f test -m rfc4716
```

Author: done

  -X POST \
  https://hub.cfengine.com/api/build/projects \
  -H 'content-type: application/json' \
  -d '
  {
  "repositoryUrl": "https://github.com/username/repository.git",
  "branch": "master",
  "authenticationType": "password",
  "password" : "git_token_or_password",
  "username" : "git_username",
  "name": "Production"
  }'
```

**Successful response example:**

```
HTTP 200 Ok
{
    "id": "8"
}
```

**Responses:**

| HTTP response code | Description |
|--------|-------------|
| 200 OK | Project successfully created |
| 422 Unprocessable entity | Validation error occurred |
| 500 Internal server error | Internal server error |

## Update project

By changing the repository url or branch you will initialize a new project and the current one will be removed from the
file system and any un-pushed/un-deployed(terminology in Mission Portal UI) changes will be lost.

**URI:** https://hub.cfengine.com/api/build/projects/:id

**Method:** PATCH

**Parameters:**

* **id** *(integer)*
  Project's ID. Required.

* **repositoryUrl** *(string)*
  Git repository URL. Project will be synchronized with this repository. Supported protocols: `http`, `https`, `ssh`

>   , `git`. Required. Git repository URL example: https://github.com/cfengine/build-index.git

Reviewer: ```suggestion
  , `git`. Required. Git repository URL example: https://github.com/username/repository.git
```
Build-index is a bad example, it's not the same format as what they would see in their own project.
